packages/mui-base/src/unstable_useNumberInput/numberInputReducer.ts

```diff
@@ -0,0 +1,154 @@
+import {
+  NumberInputActionContext,
+  NumberInputReducerAction,
+  NumberInputState,
+  StepDirection,
+} from './useNumberInput.types';
+import { NumberInputActionTypes } from './numberInputAction.types';
+import { clamp, isNumber } from './utils';
+
+// extracted from handleValueChange
+function getClampedValues(rawValue: number | undefined, context: NumberInputActionContext) {
+  const { min, max, step } = context;
+
+  const clampedValue = rawValue === undefined ? '' : clamp(rawValue, min, max, step);
+
+  const newInputValue = clampedValue === undefined ? '' : String(clampedValue);
+
+  return {
+    value: clampedValue,
+    inputValue: newInputValue,
+  };
+}
+
+function stepValue(
+  state: NumberInputState,
+  context: NumberInputActionContext,
+  direction: StepDirection,
+  multiplier: number,
+) {
+  const { value } = state;
+  const { step = 1, min, max } = context;
+
+  if (isNumber(value)) {
+    return {
+      up: value + (step ?? 1) * multiplier,
+      down: value - (step ?? 1) * multiplier,
+    }[direction];
+  }
+
+  return {
+    up: min ?? 0,
+    down: max ?? 0,
+  }[direction];
+}
+
+function handleClamp<State extends NumberInputState>(
+  state: State,
+  context: NumberInputActionContext,
+  inputValue: string,
+) {
+  const { getInputValueAsString } = context;
+
+  const parsedValue = getInputValueAsString(inputValue);
+
+  const intermediateValue =
+    parsedValue === '' || parsedValue === '-' ? undefined : parseInt(parsedValue, 10);
+
+  const clampedValues = getClampedValues(intermediateValue, context);
+
+  return {
+    ...state,
+    ...clampedValues,
+  };
+}
+
+function handleInputChange<State extends NumberInputState>(
+  state: State,
+  context: NumberInputActionContext,
+  inputValue: string,
+) {
+  const { getInputValueAsString } = context;
+
+  const parsedValue = getInputValueAsString(inputValue);
+
+  if (parsedValue === '' || parsedValue === '-') {
+    return {
+      ...state,
+      inputValue: parsedValue,
+      value: '',
+    };
+  }
+
+  if (parsedValue.match(/^-?\d+?$/)) {
+    return {
+      ...state,
+      inputValue: parsedValue,
+      value: parseInt(parsedValue, 10),
+    };
+  }
+
+  return state;
+}
+
+// use this for ArrowUp, ArrowDown, button clicks
+// use this with applyMultiplier: true for PageUp, PageDown, button shift-clicks
+function handleStep<State extends NumberInputState>(
+  state: State,
+  context: NumberInputActionContext,
+  applyMultiplier: boolean,
+  direction: StepDirection,
```

Comment:
If we used `1 | -1` as direction, we could get rid of some code without sacrificing readability IMO (using an enum would also work).

Reply:
How about `'+' | '-'` 😬

Reply:
This still requires some logic other than simple multiplication, but if you feel `1 | -1` is not expressive enough, let's leave it.